zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+- ZAP PR: [zulip/zulip-architecture#2](https://github.com/zulip/zulip-architecture/pull/2)
+
+# Summary
+
+Proposal to add end-to-end encryption for mobile push notifications.
+
+# Motivation
+
+Currently, the Zulip Push Notification Forwarding Service receives
+(but does not store or log) the contents of individual mobile push notifications.
+
+Adding end-to-end encryption for push notifications will ensure that the
+actual message content does not need to be readable by the push notification
+bouncer (or Google/Apple's servers, for that matter).
+
+# Detailed design
+
+## High-Level Plan
+
+- Client checks the server if it supports E2E encryption. If it doesn't,
+  use the legacy push notification registration flow.
+
+- Client makes an API call to the server to register itself for push notifications,
+  receives a shared secret key in response.
+
+- While sending notification, the server encrypts the notification content
+  and sends it to push bouncer. Push bouncer forwards it to FCM/APNs.
+
+- Client receives the notification via FCM/APNs and decrypts the content
+  using the shared secret key.
+
+## Data model
+
+- On the client:
+  ```
+  model Account:
+    …
+    push_account_id: RandomID[32]  # Generated by client; unique here.
+    last_acked_push_token: str  # The last push token the server acknowledged receipt of from this client.
+    push_key: SymmetricKey  # Shared key between client and server.
+  ```
+  ```
+  // A helper data structure
+  struct PushRegistration:
+    token_kind: "apns" | "fcm"
+    token: str
+    timestamp: datetime
+  ```
+- On the server, add a new table (parallel to `PushDeviceToken`):
```

Comment:
Does "parallel to" mean something concrete? Just "coexisting with", or is something more intended?

Reply:
Coexisting with seems right. The point is that we're just making a new table, not extending the existing tables we have for the legacy unencrypted flow.

Reply:
Made it:
> On the server, add a new table (coexists with `PushDeviceToken`):